text/3498-lifetime-capture-rules-2024.md

```diff
@@ -0,0 +1,743 @@
+
+When someone wants to capture a lifetime parameter not already in the bounds, that person must use one of the tricks we'll describe next.
+
+### The outlives trick
+
+Consider this example:
+
+```rust
+// error[E0700]: hidden type captures lifetime
+//               that does not appear in bounds
+fn foo<'a>(x: &'a ()) -> impl Sized { x }
+```
+
+This does not compile in Rust 2021 because the `'a` lifetime is not mentioned in the bounds of the opaque type.  We can make this work by writing:
+
+```rust
+fn foo<'a>(x: &'a ()) -> impl Sized + 'a { x }
+```
+
+This is called the "outlives trick".  But this is actually a non-solution in the general case.  Consider what `impl Sized + 'a` means.  We're returning an opaque type and promising that it outlives any lifetime `'a`.
+
+This isn't actually what we want to promise.  We want to promise that the opaque type *captures* some lifetime `'a`, and consequently, that for the opaque type to outlive some other lifetime, `'a` must outlive that other lifetime.  If we could say in Rust that a lifetime must outlive a type, we would say that the `'a` lifetime must outlive the returned opaque type.
+
+That is, the promise we're making is the wrong way around.
+
+It works anyway in this specific case only because the lifetime of the returned opaque type is *exactly* equal to the lifetime `'a`.  Because equality is symmetric, the fact that our promise is the wrong way around doesn't matter.
+
+This trick fails when there are multiple independent lifetimes that are captured, including lifetimes contained within type parameters (see [Appendix D] for an example of this).  Further, it confuses users and makes it more difficult for those users to build a consistent mental model of Rust lifetime bounds.
+
+### The `Captures` trick
+
+The correct way to express the capture of lifetime parameters in Rust 2021 is with the `Captures` trick.  It's the only option when multiple independent lifetimes must be captured (including lifetimes from captured type parameters).  Consider again our example:
+
+```rust
+// error[E0700]: hidden type captures lifetime
+//               that does not appear in bounds
+fn foo<'a>(x: &'a ()) -> impl Sized { x }
+```
+
+We could solve the problem in this way using the `Captures` trick:
+
+```rust
+trait Captures<U> {}
+impl<T: ?Sized, U> Captures<U> for T {}
+
+fn foo<'a>(x: &'a ()) -> impl Sized + Captures<&'a ()> { x }
+```
+
+Because the `'a` lifetime parameter appears in the bounds of the opaque type, Rust 2021 captures that lifetime parameter in the opaque type and accepts this code.
+
+We can extend this trick to multiple lifetimes.  For example:
+
+```rust
+fn foo<'a, 'b>(x: &'a (), y: &'b ()) -> impl Sized + Captures<(&'a (), &'b ())> {
```

Comment:
is there a reason to not use `impl Sized + Captures<'a> + Captures<'b>`?

Reply:
In the document, `Captures` is defined as:

```rust
trait Captures<U> {}
impl<T: ?Sized, U> Captures<U> for T {}
```

So it wouldn't work to say `Captures<'a>` under this definition.  Clearly we could define it differently, but in many prior discussions, people have seemed to prefer defining it in the way above.  Of course, since we never added `Captures` to the standard library (and probably never will do to this RFC), there is no one single definition of `Captures`.

Reply:
I see, I didn't even check how `Captures` was defined and I was assuming the definition we have in `rustc_data_structures`. It may be a good idea to point that out?.

Reply:
As someone who had been using this tuple spelling before, I agree that `Captures` should probably take a single lifetime parameter instead of a type parameter.

Reply:
Thanks for the pointer to how this is used internally in `rustc`.  I've added a footnote describing how this can be defined in various ways.

Reply:
Beware that a single lifetime parameter captures has its own set of pitfalls due to the fact that lifetimes do not participate in trait selection: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0f4e98df0b106b900e1c22d19b9360dc

Which is mitigated by a capture trait that is able to reference more than one lifetime: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0aa4cf0fea86aaef46e82170762aae17

Reply:
Yes, thanks @compiler-errors for mentioning that.  That's why I started using this version of `Captures` when writing examples such as in this document, as otherwise you end up having to define `Captures2`, `Captures3`, etc.  Presumably that's also why I see **nikomatsakis** always spell it this way.

I've added this caveat to the footnote.

Reply:
@compiler-errors I can not wrap my head around your example. Can you explain why the behavior is so wildly different for different definitions of `Captures`? As for someone who is not so familiar with rustc, the "lifetimes do not participate in trait selection" does not help much.